drivers/rtc/rtc_rv8263.c

```diff
@@ -0,0 +1,780 @@
+#define RV8263C8_REGISTER_MINUTES_ALARM 0x0C
+#define RV8263C8_REGISTER_HOURS_ALARM   0x0D
+#define RV8263C8_REGISTER_DATE_ALARM    0x0E
+#define RV8263C8_REGISTER_WEEKDAY_ALARM 0x0F
+#define RV8263C8_REGISTER_TIMER_VALUE   0x10
+#define RV8263C8_REGISTER_TIMER_MODE    0x11
+
+#define RV8263_BM_FAST_MODE                 (0x01 << 7)
+#define RV8263_BM_NORMAL_MODE               (0x00 << 7)
+#define RV8263C8_BM_24H_MODE_ENABLE         (0x00 << 1)
+#define RV8263C8_BM_24H_MODE_DISABLE        (0x00 << 1)
+#define RV8263C8_BM_CLOCK_ENABLE            (0x00 << 5)
+#define RV8263C8_BM_CLOCK_DISABLE           (0x01 << 5)
+#define RV8263C8_BM_ALARM_INT_ENABLE        (0x01 << 7)
+#define RV8263C8_BM_ALARM_INT_DISABLE       (0x00 << 7)
+#define RV8263C8_BM_MINUTE_INT_ENABLE       (0x01 << 5)
+#define RV8263C8_BM_MINUTE_INT_DISABLE      (0x00 << 5)
+#define RV8263C8_BM_HALF_MINUTE_INT_ENABLE  (0x01 << 4)
+#define RV8263C8_BM_HALF_MINUTE_INT_DISABLE (0x00 << 4)
+#define RV8263C8_BM_ALARM_ENABLE            (0x00 << 7)
+#define RV8263C8_BM_ALARM_DISABLE           (0x01 << 7)
+#define RV8263C8_BM_AF                      (0x01 << 6)
+#define RV8263C8_BM_TF                      (0x01 << 3)
+#define RV8263_BM_MODE                      (0x01 << 7)
+#define RV8263_BM_TD_1HZ                    (0x02 << 3)
+#define RV8263_BM_TE_ENABLE                 (0x01 << 2)
+#define RV8263_BM_TIE_ENABLE                (0x01 << 1)
+#define RV8263_BM_TI_TP_PULSE               (0x01 << 0)
+#define RV8263_BM_OS                        (0x01 << 7)
+#define RV8263C8_BM_SOFTWARE_RESET          (0x58)
+#define RV8263C8_BM_REGISTER_OFFSET         0x7F
+#define RV8263_YEAR_OFFSET                  (2000 - 1900)
+
+#define SECONDS_BITS  GENMASK(6, 0)
+#define MINUTES_BITS  GENMASK(7, 0)
+#define HOURS_BITS    GENMASK(5, 0)
+#define DATE_BITS     GENMASK(5, 0)
+#define MONTHS_BITS   GENMASK(4, 0)
+#define WEEKDAY_BITS  GENMASK(2, 0)
+#define YEAR_BITS     GENMASK(7, 0)
+#define VALIDATE_24HR BIT(6)
+
+#define MIN_SEC       0
+#define MAX_SEC       59
+#define MIN_MIN       0
+#define MAX_MIN       59
+#define MIN_HOUR      0
+#define MAX_HOUR      23
+#define MAX_WDAY      7
+#define MIN_WDAY      1
+#define MAX_MDAY      31
+#define MIN_MDAY      1
+#define MAX_MON       12
+#define MIN_MON       1
+#define MIN_YEAR_DIFF 0
+#define MAX_YEAR_DIFF 99
+
+#define DT_DRV_COMPAT microcrystal_rv_8263_c8
+
+LOG_MODULE_REGISTER(microcrystal_rv8263c8, CONFIG_RTC_LOG_LEVEL);
+
+struct rv8263c8_config {
+	struct i2c_dt_spec i2c_bus;
+	uint32_t clkout;
+
+#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(int_gpios)
+	struct gpio_dt_spec int_gpio;
+#endif
+};
+
+struct rv8263c8_data {
+	struct k_sem lock;
+
+#if (CONFIG_RTC_ALARM || CONFIG_RTC_UPDATE) && DT_ANY_INST_HAS_PROP_STATUS_OKAY(int_gpios)
+	const struct device *dev;
+#endif
+
+#if (CONFIG_RTC_ALARM || CONFIG_RTC_UPDATE) && DT_ANY_INST_HAS_PROP_STATUS_OKAY(int_gpios)
+	struct gpio_callback gpio_cb;
+#endif
+
+#if CONFIG_RTC_ALARM
```

Comment:
I think this should still be excluded if the pin is missing right? since these are all used exclusively for the callback?

Reply:
```
#if CONFIG_RTC_ALARM && DT_ANY_INST_HAS_PROP_STATUS_OKAY(int_gpios)
```

Reply:
Okay I got it. I´ve changed it back. The code is only added when the alarm is enabled via Kconfig and the device tree property `int_gpios` is available. If only the alarm is enabled without the device tree property `rv8263c8_alarm_set_callback` returns `ENOTSUP`.

Reply:
Sorry, but I´m confused now.

- `rv8263c8_alarm_set_callback` should return `ENOTSUP` when `int_gpios` is not set
- All alarm functions and the definitions from `rv8263c8_data` should be enabled when `int_gpios` is set and `CONFIG_RTC_ALARM` is set

Is this correct?

Reply:
Hi @bjarki-trackunit 

Sorry, but I´m confused now.

- `rv8263c8_alarm_set_callback` should return `ENOTSUP` when `int_gpios` is not set
- All alarm functions and the definitions from `rv8263c8_data` should be enabled when `int_gpios` is set and `CONFIG_RTC_ALARM` is set

Is this correct?

Reply:
Its just a memory optimization, the data struct does not need to contain the alarm k_work, user data and callback pointers if the alarm callback is not used, which is the case if the int gpio is missing :) The API struct was correct before :)

Reply:
The change is only required on line 107, in the data struct definition, not further down :)

Reply:
I changed it now.

- All callback-related stuff is now removed from the struct when `int_gpios` is not set and `CONFIG_RTC_ALARM` is set
- All alarm functions are included when `CONFIG_RTC_ALARM` is set only
- `ENOTSUP` is returned from `rv8263c8_alarm_set_callback` when `int_gpio` isn´t set
- The callback assignment in `rv8263c8_alarm_set_callback`  is excluded when `int_gpios` is not set to avoid warning about unused variables

```
#if DT_ANY_INST_HAS_PROP_STATUS_OKAY(int_gpios)
	data->alarm_cb = callback;
	data->alarm_cb_data = user_data;
#endif
```

Reply:
We are aligned now :)